src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -774,10 +737,26 @@ public static BigInteger DivRem(BigInteger dividend, BigInteger divisor, out Big
             if (trivialDivisor)
             {
                 uint rest;
-                uint[] bits = BigIntegerCalculator.Divide(dividend._bits, NumericsHelpers.Abs(divisor._sign), out rest);
 
-                remainder = dividend._sign < 0 ? -1 * rest : rest;
-                return new BigInteger(bits, bits, (dividend._sign < 0) ^ (divisor._sign < 0));
+                uint[]? bitsFromPool = null;
+                int size = dividend._bits.Length;
+                Span<uint> quotient = size <= BigIntegerCalculator.StackAllocThreshold ?
+                                  stackalloc uint[size]
+                                  : (bitsFromPool = ArrayPool<uint>.Shared.Rent(size)).AsSpan(0, size);
```

Comment:
I'm a bit concerned about renting from the pool and then not being able to return it since its now a span.

I wonder if there is something we can do to either account for this or to also maintain the existing path that allows a non-pooled array to still be used without copy via the `takeOwnership` functionality @stephentoub  added.

Reply:
The `try...finally` should hopefully ensure it is always returned, but that may hinder inlining or other optimizations and makes the code a bit more complicated overall

Reply:
The rented array returned back to the pool in `finally` block. The original array is saved to `bitsFromPool` variable.

Reply:
`takeOwnership` will not help. Every result is written to the rented or stack-allocated buffer. So there is no way to escape it using this parameter. The buffer passed to `BigInteger(ReadOnlySpan<uint>)` ctor must be copied using `ToArray` in any case.

Anyway, I tried to use assumption that most operations don't have leading zeroes and use heap-allocated arrays instead of renting or stack-alloc. I can conclude that this assumption is wrong because in 50% of cases the result contains leading zeroes. `BigInteger(ReadOnlySpan<uint>)` ctor in attempt to conserve space creates truncated copy of the original array. As a result, we have redundant allocation of the array for the result.

Let me explain this fact in details and choose `Add` operation for example. Before doing arithmetic operation for non-trivial operands I need to pre-allocate buffer for the result. The size of the result buffer is `Max(Digits(X), Digits(Y)) + 1` where `Digits` pseudo-function is the number of digits in the specified operand. In our case it's a length of the uints array. After computation the reserved highest position in the buffer may remain unused and must be removed from the resulting array. I see no way to predict the usage of reserved position in the result buffer and apply decision about correct allocation strategy. That's why `takeOwnership` in the current implementation not useful.

Reply:
> The try...finally should hopefully ensure it is always returned, but that may hinder inlining or other optimizations and makes the code a bit more complicated overall

Yes, but `BigIntegerCalculator.Divide` may throw `DivideByZeroException`. That's why I'm using `finally` block. I can analyze dividend before actual call of `BigIntegerCalculator.Divide` but this increase complexity.

Reply:
Right. My main concern here is additional considerations required for correctness. The new code requires a lot more ceremony and so while it does reduce allocations and improve perf, it also requires a lot more care when handling failure cases and in ensuring renting arrays are returned.

This means that future refactorings are likely going to be more error prone and may miss things. If we had a `alloca` like method, then we could create simple ref struct wrapper and do something like the following instead:
```
using (var x = alloca(size))
{
    // code
}
```

Reply:
`using` block will produce `try..finally` statements. I thought about an idea to introduce internal `ref struct` type for the purposes of Numerics library but it may cause code duplication across other .NET libraries. I'm sure that Numerics library is not the only place where such type may be useful.

Reply:
Generally speaking, I would prefer to wrap all rent/release code in `BigInteger` type to `try..finally` blocks because AFAIK `ArrayPool<uint>.Rent` may cause `OutOfMemoryException`. In `BigIntegerCalculator` it can be avoided for _leaf_ methods.

Reply:
The issue isn't really in inserting `try..finally` blocks, its that this code is very dependent on users knowing the pattern and following it correctly to ensure things work as expected.
If you don't follow the pattern, then the arrays (possibly large ones) don't get freed, leading to leaks and increased memory overhead.

So I think the gains need to be very apparent for this type of increased maintenance burden to be worthwhile. This may also be a good example of where the JIT providing an `alloca` like method would be beneficial. It's been raised in the past but hasn't happened yet.